Changes made for task 169411 (#3148)
dotnet/csharplang: spec/expressions.md

-   If T is an *enum-type*, the base types of T are the class types System.Enum, System.ValueType, and object.

> -   If T is a *struct-type*, the base types of T are the class types System.ValueType and object. 

Reviewer: additional code fences:

```suggestion
-   If `T` is a *struct-type*, the base types of `T` are the class types `System.ValueType` and `object`. 
```

Once a particular function member has been identified at binding-time, possibly through overload resolution, the actual run-time process of invoking the function member is described in §12.6.6.

> [!NOTE]

> > The following table summarizes the processing that takes place in constructs involving the six categories of function members that can be explicitly invoked. In the table, e, x, y, and value indicate expressions classified as variables or values, T indicates an expression classified as a type, F is the simple name of a method, and P is the simple name of a property.

Reviewer: additional code fences:

```suggestion
> The following table summarizes the processing that takes place in constructs involving the six categories of function members that can be explicitly invoked. In the table, `e`, `x`, `y`, and value indicate expressions classified as variables or values, `T` indicates an expression classified as a type, `F` is the simple name of a method, and `P` is the simple name of a property.
```

-   When a function member with a parameter array is invoked in its expanded form, the invocation shall specify zero or more positional arguments for the parameter array, where each argument is an expression that is implicitly convertible (§11.2) to the element type of the parameter array. In this case, the invocation creates an instance of the parameter array type with a length corresponding to the number of arguments, initializes the elements of the array instance with the given argument values, and uses the newly created array instance as the actual argument.

The expressions of an argument list are always evaluated in textual order. Thus, the example

> class Test\

Reviewer: This example should be code-fenced in the "```csharp" block.

x = 0, y = 1, z = 2\
x = 4, y = -1, z = 3

> The array co-variance rules (§17.6) permit a value of an array type A\[\] to be a reference to an instance of an array type B\[\], provided an implicit reference conversion exists from B to A. Because of these rules, when an array element of a *reference-type* is passed as a reference or output parameter, a run-time check is required to ensure that the actual element type of the array is *identical* to that of the parameter. In the following code

Reviewer: code fencing:

```suggestion
The array co-variance rules (§17.6) permit a value of an array type `A[]` to be a reference to an instance of an array type `B[]`, provided an implicit reference conversion exists from `B` to `A`. Because of these rules, when an array element of a *reference-type* is passed as a reference or output parameter, a run-time check is required to ensure that the actual element type of the array is *identical* to that of the parameter. In the following code
```

> class Test\

Reviewer: This example should be code-fenced in a "```csharp" block.

}\
}

the second invocation of F causes a System.ArrayTypeMismatchException to be thrown because the actual element type of b is string and not object. 

When a function member with a parameter array is invoked in its expanded form, the invocation is processed exactly as if an array creation expression with an array initializer (§12.7.11.5) was inserted around the expanded parameters. Given the declaration

> void F(int x, int y, params object\[\] args);

Reviewer: this should be code fenced:

```suggestion
`void F(int x, int y, params object\[\] args);`
```

F(10, 20, new object\[\] {30, 40});\

> F(10, 20, new object\[\] {1, "hello", 3.0});

Reviewer: This block (starting on line 709, and the block starting on line 703 should be in "```csharp" blocks.

In particular, note that an empty array is created when there are zero arguments given for the parameter array. 

When arguments are omitted from a function member with corresponding optional parameters, the default arguments of the function member declaration are implicitly passed. 

> [!NOTE]
> Because these are always constant, their evaluation will not impact the evaluation of the remaining arguments.

### Type inference

#### General

> When a generic method is called without specifying type arguments, a ***type inference*** process attempts to infer type arguments for the call. The presence of type inference allows a more convenient syntax to be used for calling a generic method, and allows the programmer to avoid specifying redundant type information. Given the method declaration:

Reviewer: The "example" marker should not be removed:

```suggestion
When a generic method is called without specifying type arguments, a ***type inference*** process attempts to infer type arguments for the call. The presence of type inference allows a more convenient syntax to be used for calling a generic method, and allows the programmer to avoid specifying redundant type information. [*Example*: Given the method declaration:
```

> class Chooser\

Reviewer: This should be in a code block

{\
static Random rand = new Random();

public static T Choose<T>(T first, T second) {\
return (rand.Next(2) == 0)? first: second;\
}\
}

it is possible to invoke the Choose method without explicitly specifying a type argument:

> int i = Chooser.Choose(5, 213); // Calls Choose<int>

Reviewer: These two lines should be in a code block.

If each supplied argument does not correspond to exactly one parameter in the method (§12.6.2.2), or there is a non-optional parameter with no corresponding argument, then inference immediately fails. Otherwise, assume that the generic method has the following signature:

> T~r~ M<X~1~…X~n~>(T~1~ p~1~ … T~m~ p~m~)

Reviewer: The lack of subscripts in code fences made us choose the following:

```suggestion
`Tr M<X1,...,Xn>(T1 x1, ..., Tm xm)`
```

> With a method call of the form M(E~1~ …E~m~) the task of type inference is to find unique type arguments S~1~…S~n~ for each of the type parameters X~1~…X~n~ so that the call M<S~1~…S~n~>(E~1~…E~m~)becomes valid.

Reviewer: Similar to above:

```suggestion
With a method call of the form `M(E1...Em)` the task of type inference is to find unique type arguments `S1...Sn` for each of the type parameters `X1...Xn` so that the call `M<S1...Sn>(E1...Em)` becomes valid.
```

The process of type inference is described below as an algorithm. A conformant compiler may be implemented using an alternative approach, provided it reaches the same result in all cases.

> During the process of inference each type parameter X~i~ is either *fixed* to a particular type S~i~ or *unfixed* with an associated set of *bounds.* Each of the bounds is some type T. Initially each type variable X~i~ is unfixed with an empty set of bounds.

Reviewer: The subscripts should be replaced by code fences:

```suggestion
During the process of inference each type parameter `Xi` is either *fixed* to a particular type `Si` or *unfixed* with an associated set of *bounds.* Each of the bounds is some type `T`. Initially each type variable `Xi` is unfixed with an empty set of bounds.
```

-   V is the type V~1~? and U is the type U~1~?

> -   V is a constructed type C<V~1~…V~k~> and U is a constructed type C<U~1~…U~k~>

Reviewer: These should be codefenced, not subscripts:

```suggestion
-   `V` is a constructed type `C<Vi…Vk>` and `U` is a constructed type `C<Ui…Uk>`
```

-   V is an array type V~1~\[…\]and U is an array type U~1~\[…\]of the same rank

> -   V is one of IEnumerable<V~1~>, ICollection<V~1~>, IReadOnlyList<V~1~>>, IReadOnlyCollection<V~1~> or IList<V~1~> and U is a single-dimensional array type U~1~\[\]

Reviewer: again, this should be code fenced, not subscripts. Similar edits should be made to the remaining bullet items.

```suggestion
-   `V` is one of `IEnumerable<Vi>`, `ICollection<Vi>`, `IReadOnlyList<Vi>>`, `IReadOnlyCollection<Vi>` or `IList<Vi>` and `U` is a single-dimensional array type `Ui[]`
```

-   If F is async and the body of F is either an expression classified as nothing (§12.2), or a statement block where no return statements have expressions, the inferred return type is System.Threading.Tasks.Task

> -   If F is async and has an inferred effective return type T, the inferred return type is System.Threading.Tasks.Task<T>.

Reviewer: code fence:

```suggestion
-   If `F` is async and has an inferred effective return type `T`, the inferred return type is `System.Threading.Tasks.Task<T>`.
```

-   If F is non-async and has an inferred effective return type T, the inferred return type is T.

-   Otherwise, a return type cannot be inferred for F.

> As an example of type inference involving anonymous functions, consider the Select extension method declared in the System.Linq.Enumerable class:

Reviewer: The "example" marker should not be removed:

```suggestion
-[*Example*: As an example of type inference involving anonymous functions, consider the Select extension method declared in the System.Linq.Enumerable class:
```

> namespace System.Linq\

Reviewer: This example should be in a code block.

Assuming the System.Linq namespace was imported with a using namespace directive, and given a class Customer with a Name property of type string, the Select method can be used to select the names of a list of customers:

> List<Customer> customers = GetCustomerList();\

Reviewer: These two lines should be in a code block.

IEnumerable<string> names = customers.Select(c => c.Name);

The extension method invocation (§12.7.6.3) of Select is processed by rewriting the invocation to a static method invocation:

> IEnumerable<string> names = Enumerable.Select(customers, c => c.Name);

Reviewer: Add code block here.

Since type arguments were not explicitly specified, type inference is used to infer the type arguments. First, the customers argument is related to the source parameter, inferring TSource to be Customer. Then, using the anonymous function type inference process described above, c is given type Customer, and the expression c.Name is related to the return type of the selector parameter, inferring TResult to be string. Thus, the invocation is equivalent to

> Sequence.Select<Customer,string>(customers, (Customer c) => c.Name)

Reviewer: This should be a code block.

> and the result is of type IEnumerable<string>.

Reviewer: code fence the type:

```suggestion
and the result is of type `IEnumerable<string>`.
```

The following example demonstrates how anonymous function type inference allows type information to “flow” between arguments in a generic method invocation. Given the method:

> static Z F<X,Y,Z>(X value, Func<X,Y> f1, Func<Y,Z> f2) {\

Reviewer: This should be a code block.

return f2(f1(value));\
}

Type inference for the invocation:

double seconds = F("1:15:30", s => TimeSpan.Parse(s), t => t.TotalSeconds);

> proceeds as follows: First, the argument "1:15:30" is related to the value parameter, inferring X to be string. Then, the parameter of the first anonymous function, s, is given the inferred type string, and the expression TimeSpan.Parse(s) is related to the return type of f1, inferring Y to be System.TimeSpan. Finally, the parameter of the second anonymous function, t, is given the inferred type System.TimeSpan, and the expression t.TotalSeconds is related to the return type of f2, inferring Z to be double. Thus, the result of the invocation is of type double. 

Reviewer: End example marker should not be removed:

```suggestion
proceeds as follows: First, the argument "1:15:30" is related to the value parameter, inferring X to be string. Then, the parameter of the first anonymous function, s, is given the inferred type string, and the expression TimeSpan.Parse(s) is related to the return type of f1, inferring Y to be System.TimeSpan. Finally, the parameter of the second anonymous function, t, is given the inferred type System.TimeSpan, and the expression t.TotalSeconds is related to the return type of f2, inferring Z to be double. Thus, the result of the invocation is of type double. *end example*]
```
